Ch6 Linear Model Selection and Regularization
(2) Ch6_Q9(abcde)

9. In this exercise, we will predict the number of applications received using the other variables in the College data set.<br>
***(a) Split the data set into a training set and a test set.***

In [21]:
import pandas as pd
from sklearn.model_selection import train_test_split

# 1. 加 College 資料集
college = pd.read_csv('College.csv',index_col=0)

# 2. 定義預測變數 (X) 和響應變數 (Y)
# 根據題目，'Apps' 是我們要預測的收到的申請數量。
# 其他所有變數都將作為預測變數。
X = college.drop('Apps', axis=1) # 移除 'Apps' 列作為預測變數
Y = college['Apps']              # 'Apps' 列作為響應變數

# 檢查 X 和 Y 的形狀
print(f"\n預測變數 X 的維度: {X.shape}")
print(f"響應變數 Y 的維度: {Y.shape}")

# 3. 將資料集分成訓練集和測試集
# 我們將使用 70% 的資料作為訓練集，30% 的資料作為測試集。
# random_state 參數用於確保每次運行程式碼時，分割結果都相同，以保證可重現性。
X_train, X_test, Y_train, Y_test = train_test_split(
    X, Y,
    test_size=0.3,    # 測試集佔總資料的 30%
    random_state=42   # 設定隨機種子以確保結果可重現
)

# 4. 報告訓練集和測試集的維度
print(f"\n資料分割完成:")
print(f"訓練集預測變數 (X_train) 的維度: {X_train.shape}")
print(f"訓練集響應變數 (Y_train) 的維度: {Y_train.shape}")
print(f"測試集預測變數 (X_test) 的維度: {X_test.shape}")
print(f"測試集響應變數 (Y_test) 的維度: {Y_test.shape}")

# 顯示訓練集和測試集的部分數據以供驗證
print("\n訓練集預測變數 (X_train) 的前三行:")
print(X_train.head(3))
print("\n測試集響應變數 (Y_test) 的前三行:")
print(Y_test.head(3))


預測變數 X 的維度: (777, 17)
響應變數 Y 的維度: (777,)

資料分割完成:
訓練集預測變數 (X_train) 的維度: (543, 17)
訓練集響應變數 (Y_train) 的維度: (543,)
測試集預測變數 (X_test) 的維度: (234, 17)
測試集響應變數 (Y_test) 的維度: (234,)

訓練集預測變數 (X_train) 的前三行:
    Private  Accept  Enroll  Top10perc  Top25perc  F.Undergrad  P.Undergrad  \
360     Yes    1074     397         15         40         1805          433   
336     Yes     758     428         21         46         1605          246   
110     Yes      72      51         33         71          139            3   

     Outstate  Room.Board  Books  Personal  PhD  Terminal  S.F.Ratio  \
360      9813        4050    425      1000   45        63       16.7   
336      9858        3700    450      1200   42        45       17.6   
110      8730        3600    400       800   92        92        9.3   

     perc.alumni  Expend  Grad.Rate  
360           29    7307         78  
336           16    4796         55  
110           17   10922         58  

測試集響應變數 (Y_test) 的前三行:
568    4877
530   

***(b) Fit a linear model using least squares on the training set, and report the test error obtained.***

In [30]:
import pandas as pd
from sklearn.model_selection import train_test_split
from sklearn.linear_model import LinearRegression
from sklearn.metrics import mean_squared_error
import numpy as np

# 2. 定義預測變數 (X) 和響應變數 (Y)
X = college.drop('Apps', axis=1)
Y = college['Apps']

# 由於 College 資料集中包含類別變數 (Private)，我們需要進行獨熱編碼 (One-Hot Encoding)
# LinearRegression 期望所有輸入都是數值型
X = pd.get_dummies(X, drop_first=True) # drop_first=True 避免多重共線性陷阱

# 3. 將資料集分成訓練集和測試集
X_train, X_test, Y_train, Y_test = train_test_split(
    X, Y,
    test_size=0.3,
    random_state=42
)

# --- (b) 在訓練集上擬合線性模型並報告測試誤差 ---

# 1. 創建線性迴歸模型實例
linear_model = LinearRegression()

# 2. 在訓練集上擬合模型
linear_model.fit(X_train, Y_train)

# 3. 使用擬合的模型在測試集上進行預測
Y_pred_test = linear_model.predict(X_test)

# 4. 計算測試集上的均方誤差 (MSE)
test_mse = mean_squared_error(Y_test, Y_pred_test)

# 5. 報告測試誤差
print(f"\n--- 線性迴歸模型結果 ---")
print(f"測試集上的均方誤差 (MSE): {test_mse:.2f}")

# 額外資訊：顯示模型係數
print("\n模型係數:")
# 結合特徵名稱和係數
coefficients = pd.Series(linear_model.coef_, index=X_train.columns)
print(coefficients)
print(f"\n截距 (Intercept): {linear_model.intercept_:.2f}")

# 也可以報告 R-squared，但題目只要求 MSE
# train_r2 = linear_model.score(X_train, Y_train)
# test_r2 = linear_model.score(X_test, Y_test)
# print(f"訓練集 R-squared: {train_r2:.4f}")
# print(f"測試集 R-squared: {test_r2:.4f}")


--- 線性迴歸模型結果 ---
測試集上的均方誤差 (MSE): 1931803.19

模型係數:
Accept           1.752032
Enroll          -1.314120
Top10perc       32.726179
Top25perc       -3.669620
F.Undergrad      0.035782
P.Undergrad      0.020205
Outstate        -0.079454
Room.Board       0.137947
Books            0.189046
Personal         0.080150
PhD             -9.352419
Terminal        -1.106691
S.F.Ratio       19.327165
perc.alumni      1.419874
Expend           0.062600
Grad.Rate        7.649710
Private_Yes   -557.508542
dtype: float64

截距 (Intercept): -622.25


In [32]:
X

,Accept,Enroll,Top10perc,Top25perc,F.Undergrad,P.Undergrad,Outstate,Room.Board,Books,Personal,PhD,Terminal,S.F.Ratio,perc.alumni,Expend,Grad.Rate,Private_Yes
0,1232,721,23,52,2885,537,7440,3300,450,2200,70,78,18.1,12,7041,60,True
1,1924,512,16,29,2683,1227,12280,6450,750,1500,29,30,12.2,16,10527,56,True
2,1097,336,22,50,1036,99,11250,3750,400,1165,53,66,12.9,30,8735,54,True
3,349,137,60,89,510,63,12960,5450,450,875,92,97,7.7,37,19016,59,True
4,146,55,16,44,249,869,7560,4120,800,1500,76,72,11.9,2,10922,15,True
...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...
772,1515,543,4,26,3089,2029,6797,3900,500,1200,60,60,21.0,14,4469,40,False
773,1805,695,24,47,2849,1107,11520,4960,600,1250,73,75,13.3,31,9189,83,True
774,1915,695,34,61,2793,166,6900,4200,617,781,67,75,14.4,20,8323,49,True
775,2453,1317,95,99,5217,83,19840,6510,630,2115,96,96,5.8,49,40386,99,True


In [34]:
# 也可以報告 R-squared，但題目只要求 MSE
train_r2 = linear_model.score(X_train, Y_train)
test_r2 = linear_model.score(X_test, Y_test)
print(f"訓練集 R-squared: {train_r2:.4f}")
print(f"測試集 R-squared: {test_r2:.4f}")

訓練集 R-squared: 0.9482
測試集 R-squared: 0.8573


***(c) Fit a ridge regression model on the training set, with λ chosen by cross-validation. Report the test error obtained.***

In [39]:
from sklearn.linear_model import LinearRegression, RidgeCV
from sklearn.preprocessing import StandardScaler
from sklearn.metrics import mean_squared_error
import numpy as np
# --- (c) 擬合 Ridge Regression 模型並報告測試誤差 ---

# 1. 標準化預測變數
# Ridge Regression 對變數尺度敏感，因此需要進行標準化
scaler = StandardScaler()
X_train_scaled = scaler.fit_transform(X_train) # 在訓練集上擬合 StandardScaler 並轉換
X_test_scaled = scaler.transform(X_test)       # 使用訓練集的 StandardScaler 轉換測試集

# 將標準化後的數據轉換回 DataFrame，保留列名，以便模型解釋
X_train_scaled_df = pd.DataFrame(X_train_scaled, columns=X_train.columns)
X_test_scaled_df = pd.DataFrame(X_test_scaled, columns=X_test.columns)


# 2. 創建 RidgeCV 模型實例並使用交叉驗證選擇最佳 alpha (λ)
# alphas 參數用於指定要測試的 λ 值範圍
# 我們通常在對數尺度上選擇一個範圍，因為 λ 的影響是非線性的
alphas = np.logspace(-4, 4, 200) # 測試從 0.0001 到 10000 的 200 個 alpha 值
                                 # 根據數據情況可以調整範圍

ridge_cv_model = RidgeCV(
    alphas=alphas,
    cv=10,             # 使用 10 折交叉驗證
    scoring='neg_mean_squared_error' # 選擇最小化 MSE 的 alpha
)

# 3. 在標準化後的訓練集上擬合模型
ridge_cv_model.fit(X_train_scaled_df, Y_train)

# 4. 獲取由交叉驗證選出的最佳 alpha (λ)
optimal_alpha_ridge = ridge_cv_model.alpha_

# 5. 使用最佳模型在標準化後的測試集上進行預測
Y_pred_test_ridge = ridge_cv_model.predict(X_test_scaled_df)

# 6. 計算測試集上的均方誤差 (MSE)
test_mse_ridge = mean_squared_error(Y_test, Y_pred_test_ridge)

# 7. 報告結果
print(f"\n--- 嶺迴歸 (Ridge Regression) 模型結果 ---")
print(f"通過交叉驗證選出的最佳 λ (alpha): {optimal_alpha_ridge:.6f}")
print(f"測試集上的均方誤差 (MSE): {test_mse_ridge:.2f}")

# 額外資訊：顯示最佳 Ridge 模型係數
print("\n最佳 Ridge 模型係數 (標準化數據):")
# 獲取最佳模型的係數。ridge_cv_model 已經是擬合好的最佳模型。
ridge_coefficients = pd.Series(ridge_cv_model.coef_, index=X_train.columns)
print(ridge_coefficients)
print(f"\n截距 (Intercept): {ridge_cv_model.intercept_:.2f}")

# --- (b) 線性迴歸模型結果 (重複報告用於比較) ---
print(f"\n--- (b) 線性迴歸模型結果 (用於比較) ---")
linear_model = LinearRegression()
linear_model.fit(X_train, Y_train)
Y_pred_test_linear = linear_model.predict(X_test)
test_mse_linear = mean_squared_error(Y_test, Y_pred_test_linear)
print(f"測試集上的均方誤差 (MSE): {test_mse_linear:.2f}")


--- 嶺迴歸 (Ridge Regression) 模型結果 ---
通過交叉驗證選出的最佳 λ (alpha): 0.000100
測試集上的均方誤差 (MSE): 1931799.84

最佳 Ridge 模型係數 (標準化數據):
Accept         4474.737012
Enroll        -1158.039715
Top10perc       559.587784
Top25perc       -70.715697
F.Undergrad     163.158141
P.Undergrad      26.736756
Outstate       -320.394176
Room.Board      152.779636
Books            30.551882
Personal         53.258643
PhD            -151.037922
Terminal        -16.330832
S.F.Ratio        77.801253
perc.alumni      17.092783
Expend          323.985721
Grad.Rate       132.262593
Private_Yes    -248.245041
dtype: float64

截距 (Intercept): 2964.50

--- (b) 線性迴歸模型結果 (用於比較) ---
測試集上的均方誤差 (MSE): 1931803.19


### 結果報告與討論：

**情境**：您正在使用 College 資料集，目標是預測 Apps (收到的申請數量)，並將資料分成了訓練集和測試集。

---

1.  **最佳 $\lambda$ (alpha) 值**：
    您通過 10 折交叉驗證，RidgeCV 選擇出的最佳正則化參數 $\lambda$ (alpha) 為 `0.000100`。
    *   **解釋**：這個值是 `RidgeCV` 在您設定的 `alphas` 範圍內（通常是從極小到極大的對數尺度值），通過交叉驗證找到了能夠最小化平均測試誤差（在此處是負均方誤差）的 $\lambda$ 值。
    *   **意義**：`0.0001` 是一個**非常小**的 $\lambda$ 值。這意味著模型對係數的 $L_2$ 懲罰作用**極其微弱**。當 $\lambda$ 值接近於零時，嶺迴歸的行為會非常接近於普通的最小平方法 (OLS)。

2.  **測試誤差 (Test Error)**：
    嶺迴歸模型在測試集上獲得的均方誤差 (MSE) 為 `1931799.84`。
    *   **解釋**：這個數值量化了模型預測的 Apps 值與實際 Apps 值之間的平均平方差異。

3.  **與線性迴歸 (OLS) 的比較**：
    *   您報告的線性迴歸模型 (OLS) 的測試 MSE 為 `1931803.19`。
    *   **比較**：嶺迴歸模型的測試 MSE (`1931799.84`) 略低於線性迴歸模型 (`1931803.19`)。兩者之間的差異僅為約 `3.35`。
    *   **討論**：這個非常微小的差異表明，對於您當前的數據分割和資料特性而言，嶺迴歸所引入的正則化幾乎沒有帶來顯著的預測性能提升。這與最佳 $\lambda$ 值極小（接近 OLS）的觀察結果是一致的。

4.  **模型係數 (標準化數據)**：
    您的報告列出了所有預測變數的係數，以及截距。
    *   **係數特性**：由於嶺迴歸的特性，所有係數都是非零的。然而，因為最佳 $\lambda$ 值非常小，這些係數的縮減程度相對較小，與 OLS 的係數可能相差不大。
    *   **係數大小**：`Accept` (`4474.74`) 和 `Enroll` (`-1158.04`) 的係數絕對值相對較大，這表明這兩個變數對 Apps 的影響力較大。`Private_Yes` 的係數為 `-248.25`，表示私立大學（相對於非私立大學）在其他條件相同的情況下，平均申請數量較少。
    *   **截距**：`2964.50`。由於係數是在標準化後的數據上計算的，這個截距值代表當所有標準化後的預測變數都為零時（即原始預測變數都取其平均值時）Apps 的預測值。

### **總結與結論：**

*   **嶺迴歸的邊際效用**：在您這次的實驗中，嶺迴歸相較於簡單的最小平方法 (OLS) 而言，幾乎沒有提供額外的預測性能優勢。其最佳 $\lambda$ 值極小，這暗示著模型在沒有強烈懲罰的情況下就能表現良好，或者數據集本身（或特定的訓練/測試分割）並未呈現出讓嶺迴歸顯著優於 OLS 的強烈多重共線性或過擬合問題。
*   **數據特性**：這可能說明 College 資料集中的預測變數之間雖然可能存在一定程度的相關性，但這種相關性對於影響 OLS 模型的預測能力而言，不足以產生嚴重的負面影響。此外，數據的樣本量可能也足夠大，使得 OLS 估計相對穩定。
*   **比較分析**：這個結果是模型評估中一個重要的發現。它告訴我們，並非所有時候引入更複雜的模型（如帶有正則化的模型）都會帶來性能的顯著提升。在某些情況下，簡單的 OLS 模型可能就已經足夠好。這也為後續的 Lasso 模型提供了一個基準，看看 Lasso 在這種數據下能否帶來不同的結果。

***(d) Fit a lasso model on the training set, with λ chosen by cross-validation. Report the test error obtained, along with the number of non-zero coefficient estimates.***

In [43]:
from sklearn.linear_model import LinearRegression, RidgeCV, LassoCV
from sklearn.preprocessing import StandardScaler
from sklearn.metrics import mean_squared_error
import numpy as np
import warnings

# 忽略來自 sklearn 可能關於收斂的警告
warnings.filterwarnings('ignore')

# --- (d) 擬合 Lasso 模型並報告測試誤差及非零係數數量 ---

# 1. 創建 LassoCV 模型實例並使用交叉驗證選擇最佳 alpha (λ)
# alphas 參數用於指定要測試的 λ 值範圍
# 我們通常在對數尺度上選擇一個範圍，因為 λ 的影響是非線性的
# max_iter 增加迭代次數以確保收斂
alphas_lasso = np.logspace(-4, 0, 100) # 測試從 0.0001 到 1 的 100 個 alpha 值
                                     # 通常 Lasso 的 alpha 範圍需要比 Ridge 更廣泛，且最大值不能太大，否則所有係數都為零

lasso_cv_model = LassoCV(
    alphas=alphas_lasso,
    cv=10,                      # 使用 10 折交叉驗證
    random_state=42,            # 確保交叉驗證分割的可重現性
    n_jobs=-1,                  # 使用所有可用處理器進行並行計算
    max_iter=10000              # 增加迭代次數以幫助收斂
)

# 2. 在標準化後的訓練集上擬合模型
lasso_cv_model.fit(X_train_scaled_df, Y_train)

# 3. 獲取由交叉驗證選出的最佳 alpha (λ)
optimal_alpha_lasso = lasso_cv_model.alpha_

# 4. 使用最佳模型在標準化後的測試集上進行預測
Y_pred_test_lasso = lasso_cv_model.predict(X_test_scaled_df)

# 5. 計算測試集上的均方誤差 (MSE)
test_mse_lasso = mean_squared_error(Y_test, Y_pred_test_lasso)

# 6. 報告結果
print(f"\n--- Lasso 模型結果 ---")
print(f"通過交叉驗證選出的最佳 λ (alpha): {optimal_alpha_lasso:.6f}")
print(f"測試集上的均方誤差 (MSE): {test_mse_lasso:.2f}")

# 7. 報告非零係數的數量
lasso_coefficients = lasso_cv_model.coef_
non_zero_coefficients_count = np.sum(lasso_coefficients != 0)
print(f"非零係數估計的數量: {non_zero_coefficients_count}")

# 額外資訊：顯示非零係數
print("\n非零係數估計 (標準化數據):")
coefficients_series = pd.Series(lasso_coefficients, index=X_train.columns)
for feature, coef in coefficients_series.items():
    if coef != 0:
        print(f"  {feature}: {coef:.4f}")
print(f"\n截距 (Intercept): {lasso_cv_model.intercept_:.2f}")


# --- 報告 (b) 和 (c) 的結果用於比較 ---
print(f"\n--- 各模型測試誤差比較 ---")

# (b) 線性迴歸模型結果
linear_model = LinearRegression()
linear_model.fit(X_train, Y_train)
Y_pred_test_linear = linear_model.predict(X_test)
test_mse_linear = mean_squared_error(Y_test, Y_pred_test_linear)
print(f"線性迴歸 (OLS) 測試 MSE: {test_mse_linear:.2f}")

# (c) 嶺迴歸模型結果
# 需要重新執行 RidgeCV 才能獲取其 MSE
alphas_ridge = np.logspace(-4, 4, 200)
ridge_cv_model_compare = RidgeCV(
    alphas=alphas_ridge,
    cv=10,
    scoring='neg_mean_squared_error'
)
ridge_cv_model_compare.fit(X_train_scaled_df, Y_train)
Y_pred_test_ridge = ridge_cv_model_compare.predict(X_test_scaled_df)
test_mse_ridge = mean_squared_error(Y_test, Y_pred_test_ridge)
print(f"嶺迴歸 (Ridge) 測試 MSE: {test_mse_ridge:.2f}")

print(f"Lasso 測試 MSE: {test_mse_lasso:.2f}")



--- Lasso 模型結果 ---
通過交叉驗證選出的最佳 λ (alpha): 0.000100
測試集上的均方誤差 (MSE): 1931803.39
非零係數估計的數量: 17

非零係數估計 (標準化數據):
  Accept: 4474.7438
  Enroll: -1158.0466
  Top10perc: 559.5882
  Top25perc: -70.7158
  F.Undergrad: 163.1584
  P.Undergrad: 26.7369
  Outstate: -320.3942
  Room.Board: 152.7789
  Books: 30.5518
  Personal: 53.2588
  PhD: -151.0380
  Terminal: -16.3306
  S.F.Ratio: 77.8010
  perc.alumni: 17.0929
  Expend: 323.9854
  Grad.Rate: 132.2621
  Private_Yes: -248.2452

截距 (Intercept): 2964.50

--- 各模型測試誤差比較 ---
線性迴歸 (OLS) 測試 MSE: 1931803.19
嶺迴歸 (Ridge) 測試 MSE: 1931799.84
Lasso 測試 MSE: 1931803.39


#### **詳細解釋與討論：**

1.  **最佳 $\lambda$ (alpha) 值**：
    *   **數值**：您通過 10 折交叉驗證選擇出的最佳正則化參數 $\lambda$ (alpha) 約為 `0.000100`。
    *   **意義**：這是一個**非常小**的 $\lambda$ 值。在 Lasso 回歸中，$\lambda$ 控制著懲罰項的強度。當 $\lambda$ 值接近於零時，Lasso 的懲罰作用極弱，其行為會非常接近於普通的最小平方法 (OLS)。

2.  **測試誤差 (Test Error)**：
    *   **數值**：Lasso 模型在測試集上獲得的均方誤差 (MSE) 為 `1931803.39`。這個數值量化了模型預測的 `Apps` 值與實際 `Apps` 值之間的平均平方差異。

3.  **非零係數估計的數量**：
    *   **數值**：Lasso 模型中，非零係數估計的數量為 `17`。
    *   **意義**：這表示**所有 17 個預測變數**（包括 `Private_Yes` 獨熱編碼後）都被 Lasso 保留在模型中，沒有任何係數被完全壓縮為零。

4.  **模型係數 (標準化數據)**：
    *   **觀察**：報告中列出了所有 17 個非零係數的估計值，以及截距 `2964.50`。由於所有變數都被保留，這些係數與在微弱懲罰下的 OLS 係數非常相似。係數的絕對值大小反映了在標準化數據中該變數對 `Apps` 的相對重要性。例如，`Accept` 具有最大的正係數，`Enroll` 具有最大的負係數。

#### **與其他模型測試誤差的比較：**

*   **線性迴歸 (OLS) 測試 MSE**：`1931803.19`
*   **嶺迴歸 (Ridge) 測試 MSE**：`1931799.84`
*   **Lasso 測試 MSE**：`1931803.39`

*   **比較分析**：
    *   在您這次的實驗中，所有三個模型（OLS、Ridge、Lasso）在測試集上的 MSE 值都非常接近。Lasso 的 MSE 甚至略高於 OLS 和 Ridge。
    *   這再次強烈暗示在這個特定的 `College` 資料集和您所做的數據分割下，**模型的過擬合問題並不嚴重**。因此，正則化模型（Ridge 和 Lasso）的優勢未能充分體現，它們未能帶來顯著的預測性能提升。

#### **Lasso 在本案例中的表現總結：**

*   **未能實現特徵選擇**：Lasso 的一個主要優勢是能夠通過將不重要變數的係數壓縮為零來執行特徵選擇。然而，在您的結果中，**Lasso 並未將任何係數壓縮為零**，所有 17 個預測變數都被保留在模型中。
*   **原因分析**：
    1.  **極小的 $\lambda$ 值**：交叉驗證選擇的最佳 $\lambda$ 值 `0.000100` 非常小。這麼小的懲罰強度意味著模型幾乎沒有被要求犧牲任何係數來換取模型簡潔性。如果將任何係數壓縮為零會導致預測誤差增加，那麼 `LassoCV` 為了最小化預測誤差，就會傾向於不進行特徵選擇。
    2.  **所有變數可能都有貢獻**：這可能表明 `College` 資料集中的所有預測變數都對 `Apps` 的預測具有一定的貢獻，或者它們之間存在足夠的共線性，使得 Lasso 難以將任何一個變數的係數單獨壓縮為零而不顯著增加模型誤差。
    3.  **過擬合不嚴重**：所有模型預測性能的相似性也支持了這一點：由於沒有明顯的過擬合問題，正則化方法（Lasso）在性能提升和特徵選擇方面也就沒有太多發揮的空間。

**總體結論**：

在這個 `College` 資料集的特定情況下，Lasso 模型未能發揮其核心的特徵選擇功能。儘管它找到了最佳的 $\lambda$ 值，但這個值非常小，導致模型行為類似於 OLS，保留了所有預測變數。這說明正則化方法的效果是數據依賴的；並非所有數據集都會從強力的正則化或特徵選擇中獲得顯著的預測性能提升。當資料本身的過擬合程度不高或變數之間相關性結構不強烈需要懲罰時，Lasso 的特徵選擇優勢可能不會顯現。


***(e) Fit a PCR model on the training set, with M chosen by cross- validation. Report the test error obtained, along with the value of M selected by cross-validation.***

In [47]:
from sklearn.model_selection import train_test_split, KFold, cross_val_score
from sklearn.linear_model import LinearRegression, RidgeCV, LassoCV
from sklearn.preprocessing import StandardScaler
from sklearn.decomposition import PCA
from sklearn.metrics import mean_squared_error
import numpy as np
import warnings

# 忽略來自 sklearn 可能關於收斂的警告
warnings.filterwarnings('ignore')

# --- (e) 擬合 PCR 模型並報告測試誤差及 M 值 ---

# 1. 對訓練集數據執行 PCA
# 我們先不指定 n_components，讓 PCA 計算所有可能的主成分
pca = PCA()
pca.fit(X_train_scaled_df)

# 將訓練集數據轉換為主成分
# X_train_pca 的列數會等於原始特徵的數量 (17個)
X_train_pca = pca.transform(X_train_scaled_df)

# 2. 使用交叉驗證選擇最佳的 M (主成分數量)
# 我們將嘗試使用 M = 1 到 原始特徵數 (17) 的所有主成分
n_components = X_train_scaled_df.shape[1] # 原始特徵數量

# 儲存每個 M 值對應的交叉驗證 MSE
cv_mse_scores = []

# 設置 KFold 交叉驗證
kf = KFold(n_splits=10, shuffle=True, random_state=42)

for M in range(1, n_components + 1):
    # 使用前 M 個主成分
    X_train_components = X_train_pca[:, :M]

    # 擬合線性迴歸模型並進行交叉驗證
    # scoring='neg_mean_squared_error' 返回負 MSE，所以我們取負數來得到 MSE
    model = LinearRegression()
    mse_scores = -cross_val_score(model, X_train_components, Y_train,
                                   cv=kf, scoring='neg_mean_squared_error')
    cv_mse_scores.append(np.mean(mse_scores))

# 找到最小 CV MSE 對應的 M 值
optimal_M = np.argmin(cv_mse_scores) + 1 # +1 因為 M 是從 1 開始計數

print(f"\n--- 主成分迴歸 (PCR) 模型結果 ---")
print(f"通過交叉驗證選出的最佳主成分數量 M: {optimal_M}")

# 3. 使用最佳 M 值重新擬合 PCR 模型
# 將訓練集轉換為最佳 M 個主成分
X_train_final_components = X_train_pca[:, :optimal_M]
final_pcr_model = LinearRegression()
final_pcr_model.fit(X_train_final_components, Y_train)

# 4. 將測試集轉換為相同的主成分空間，並使用相同的 M
# 注意：使用訓練集上擬合的 pca 對測試集進行轉換
X_test_pca = pca.transform(X_test_scaled)
X_test_final_components = X_test_pca[:, :optimal_M]

# 5. 在測試集上進行預測
Y_pred_test_pcr = final_pcr_model.predict(X_test_final_components)

# 6. 計算測試集上的均方誤差 (MSE)
test_mse_pcr = mean_squared_error(Y_test, Y_pred_test_pcr)

# 7. 報告測試誤差
print(f"測試集上的均方誤差 (MSE): {test_mse_pcr:.2f}")


# --- 報告 (b), (c), (d) 的結果用於比較 ---
print(f"\n--- 各模型測試誤差比較 ---")

# (b) 線性迴歸模型結果
linear_model = LinearRegression()
linear_model.fit(X_train, Y_train)
Y_pred_test_linear = linear_model.predict(X_test)
test_mse_linear = mean_squared_error(Y_test, Y_pred_test_linear)
print(f"線性迴歸 (OLS) 測試 MSE: {test_mse_linear:.2f}")

# (c) 嶺迴歸模型結果
alphas_ridge = np.logspace(-4, 4, 200)
ridge_cv_model_compare = RidgeCV(
    alphas=alphas_ridge,
    cv=10,
    scoring='neg_mean_squared_error'
)
ridge_cv_model_compare.fit(X_train_scaled_df, Y_train)
Y_pred_test_ridge = ridge_cv_model_compare.predict(X_test_scaled_df)
test_mse_ridge = mean_squared_error(Y_test, Y_pred_test_ridge)
print(f"嶺迴歸 (Ridge) 測試 MSE: {test_mse_ridge:.2f}")

# (d) Lasso 模型結果
alphas_lasso = np.logspace(-4, 0, 100)
lasso_cv_model_compare = LassoCV(
    alphas=alphas_lasso,
    cv=10,
    random_state=42,
    n_jobs=-1,
    max_iter=10000
)
lasso_cv_model_compare.fit(X_train_scaled_df, Y_train)
Y_pred_test_lasso = lasso_cv_model_compare.predict(X_test_scaled_df)
test_mse_lasso = mean_squared_error(Y_test, Y_pred_test_lasso)
print(f"Lasso 測試 MSE: {test_mse_lasso:.2f}")

print(f"主成分迴歸 (PCR) 測試 MSE: {test_mse_pcr:.2f}")



--- 主成分迴歸 (PCR) 模型結果 ---
通過交叉驗證選出的最佳主成分數量 M: 17
測試集上的均方誤差 (MSE): 1931803.19

--- 各模型測試誤差比較 ---
線性迴歸 (OLS) 測試 MSE: 1931803.19
嶺迴歸 (Ridge) 測試 MSE: 1931799.84
Lasso 測試 MSE: 1931803.39
主成分迴歸 (PCR) 測試 MSE: 1931803.19


#### **詳細解釋與討論：**

1.  **最佳 M 值 (主成分數量)**：
    *   **數值**：您通過交叉驗證選出的最佳主成分數量 M 為 **`17`**。
    *   **意義**：原始數據集 `X` 中，經過獨熱編碼後共有 `17` 個預測變數（`Private` 變成 `Private_Yes`，加上其餘 16 個數值變數）。當 PCR 選擇的最佳 M 值等於原始預測變數的總數時，這意味著主成分分析 (PCA) 實際上**沒有執行任何維度縮減**。模型使用了所有經 PCA 轉換後的成分來擬合迴歸。

2.  **PCR 測試誤差 (Test Error)**：
    *   **數值**：PCR 模型在測試集上獲得的均方誤差 (MSE) 為 **`1931803.19`**。

3.  **與其他模型的比較**：
    *   **與 OLS 的關係**：
        您會注意到 PCR 的測試 MSE (`1931803.19`) **與線性迴歸 (OLS) 的測試 MSE (`1931803.19`) 完全相同**。這並非巧合。當 PCR 使用所有可能的主成分 (M 等於原始特徵數量) 時，由於主成分是原始特徵的線性組合，並且它們是正交的，那麼使用所有主成分擬合的線性迴歸模型在數學上將等價於在原始（且經過標準化）特徵上擬合的 OLS 模型。也就是說，當 M 等於特徵總數時，PCR 實質上就是 OLS。

    *   **與 Ridge 和 Lasso 的關係**：
        *   Ridge 測試 MSE: `1931799.84`
        *   Lasso 測試 MSE: `1931803.39`
        *   Lasso 的 MSE 略高於 OLS 和 PCR，而 Ridge 的 MSE 略低於 OLS 和 PCR。所有四個模型在測試集上的 MSE 值都**非常接近**。

#### **本案例中 PCR 的表現總結與討論：**

*   **M=17 的意義**：交叉驗證選擇 M=17 表示在這個特定的數據分割中，保留所有主成分對預測 `Apps` 最為有利。換句話說，如果捨棄任何一個主成分（即使是貢獻率最小的），都會導致模型在交叉驗證中表現變差。這暗示了即使是「較不重要」的主成分，也可能攜帶了對響應變數有用的信息。
*   **預測性能**：由於 M=17 使得 PCR 退化為 OLS，因此其預測性能與 OLS 完全一致。這也解釋了為何其 MSE 值與 OLS 模型的 MSE 相同。
*   **模型行為**：
    *   PCR 的目的是通過選擇較少數量的 `M < p` 個主成分來減少維度，以期降低方差並提高預測性能，尤其是在處理多重共線性時。
    *   然而，當最佳 M 值被選定為等於所有原始特徵數量時，它暗示了在當前資料集中，預測變數之間的共線性問題可能不夠嚴重，或者說，即使存在共線性，它對 OLS 模型的預測精度也沒有造成足夠的負面影響，以至於需要通過維度縮減來改善。
    *   此外，所有四個模型（OLS、Ridge、Lasso、PCR）的測試 MSE 都極其接近，這是一個強烈的信號，表明在這個 `College` 資料集上，模型**沒有明顯的過擬合問題**，並且預測變數之間**也沒有嚴重的多重共線性問題**，以至於需要正則化或降維技術來大幅度改善預測性能。

**總體結論**：

在您這次的實驗中，主成分迴歸 (PCR) 選擇使用全部 17 個主成分，這使其在效果上等同於在標準化數據上進行 OLS 迴歸。其測試誤差與 OLS 幾乎相同，並且與經過輕微正則化的 Ridge 和 Lasso 模型也相差無幾。這個結果表明，對於這個 `College` 資料集和特定的訓練/測試分割，原始的線性迴歸模型 (OLS) 已經表現得相當好，沒有明顯的過擬合或多重共線性問題需要更複雜的降維或正則化技術來解決。